moebot_bot/util/db/types/role.go

```diff
@@ -0,0 +1,33 @@
+package types
+
+import "database/sql"
+
+// Permission enum
+type Permission int
+
+const RoleCodeSearchText = "[code]"
+const RoleCodeLength = 6
+
+const (
+	// Default permission level, no permissions regarding what can or can't be done
+	PermAll Permission = 2
+	// Mod level permission, allowed to do some server changing commands
+	PermMod Permission = 50
+	// Guild Owner permission. Essentially a master
+	PermGuildOwner Permission = 90
+	// Used to disable something, no one can have this permission level
+	PermNone Permission = 100
+	// Master level permission, can't ever be ignored or disabled
+	PermMaster Permission = 101
+)
+
+type Role struct {
+	Id                         int
+	ServerId                   int
+	Groups                     []int
```

Comment:
The structs for the tables should always be a 1-to-1 mapping with the database tables so we'll need to move this to a separate struct. Perhaps a single struct with a role + list of groups (all their info instead of ints) called `RolesWithGroups` or something similar (I'm not a huge fan of that name tbh)

Reply:
While a perfectly logical suggestion, this conflicts with a lot of existing logic, namely all the functions checking if a group is related to a role, and role adding/removing in general.

Go has no slice utilities of its own, and we'd have to write our own helper functions to do contains/subtract/index with slices of groups. 

I'd rather keep it as a slice of group Ids and having to work with slices of int rather than transforming it to a slice of groups and working with slices of groups or (even worse) pointers of groups.

Reply:
That's fine to leave it as a slice of ints rather than groups, that part was just a suggestion to possibly ease the changes. 

My point was more for the first part: that the `Role` struct _must_ match the db table perfectly. This is going to be strictly enforced with the upcoming db change, meaning this struct would break and it would have to be refactored anyways. I'd much rather do that now than save it for later, especially when you're working on this right now.

Reply:
After looking at the ORM we're gonna introduced, as already discussed, I think we should do the refactoring when we actually implement it, since we'll have to touch many-to-many relationships anyway.